internal/runners/builds/builds.go

```diff
@@ -192,3 +187,72 @@ func (b *Builds) outputPlain(out *StructuredOutput, fullID bool) error {
 	b.out.Print("\nTo download builds run '[ACTIONABLE]state builds dl <ID>[/RESET]'.")
 	return nil
 }
+
+// getTerminalArtifactMap returns a project's terminal artifact map, depending on the given
+// arguments. By default, the map for the current project is returned, but a map for a given
+// commitID for the current project can be returned, as can the map for a remote project
+// (and optional commitID).
+func getTerminalArtifactMap(
+	pj *project.Project,
+	namespace *project.Namespaced,
+	commit string,
+	auth *authentication.Auth,
+	an analytics.Dispatcher,
+	svcModel *model.SvcModel,
+	out output.Outputer,
+	cfg model.Configurable) (buildplan.TerminalArtifactMap, error) {
+	if pj == nil && !namespace.IsValid() {
+		return nil, rationalize.ErrNoProject
+	}
+
+	commitID := strfmt.UUID(commit)
+	if commitID != "" && !strfmt.IsUUID(commitID.String()) {
+		return nil, locale.NewInputError("err_commit_id_invalid", "", commitID.String())
+	}
+
+	if !namespace.IsValid() {
+		if commitID != "" {
+			// Return artifact map from the given commitID for the current project.
+			bp := model.NewBuildPlannerModel(auth)
+			buildPlan, err := bp.FetchBuildResult(commitID, pj.Owner(), pj.Name())
+			if err != nil {
+				return nil, errs.Wrap(err, "Failed to fetch build plan")
+			}
+			return buildplan.NewMapFromBuildPlan(buildPlan.Build, false, false, nil)
+		}
+
+		// Return the artifact map from this runtime.
+		rt, err := runtime.NewFromProject(pj, target.TriggerBuilds, an, svcModel, out, auth, cfg)
+		if err != nil {
+			return nil, locale.WrapInputError(err, "err_refresh_runtime_new", "Could not update runtime for this project.")
+		}
+		return rt.TerminalArtifactMap(false)
+	}
+
+	if commitID == "" {
+		// Return the artifact map for the latest commitID of the given project.
+		pj, err := model.FetchProjectByName(namespace.Owner, namespace.Project)
+		if err != nil {
+			return nil, locale.WrapInputError(err, "err_fetch_project", "", namespace.String())
+		}
+
+		branch, err := model.DefaultBranchForProject(pj)
+		if err != nil {
+			return nil, errs.Wrap(err, "Could not grab branch for project")
+		}
+
+		commitUUID, err := model.BranchCommitID(namespace.Owner, namespace.Project, branch.Label)
+		if err != nil {
+			return nil, errs.Wrap(err, "Could not get commit ID for project")
+		}
+		commitID = *commitUUID
+	}
+
+	// Return the artifact map for the given commitID of the given project.
+	bp := model.NewBuildPlannerModel(auth)
+	buildPlan, err := bp.FetchBuildResult(commitID, namespace.Owner, namespace.Project)
+	if err != nil {
+		return nil, errs.Wrap(err, "Failed to fetch build plan")
+	}
+	return buildplan.NewMapFromBuildPlan(buildPlan.Build, false, false, nil)
```

Comment:
I see 3 behaviours here, but I think it's hard to grok that from this code. Can we make this a switch case where each case holds the behaviour?

So

- Source from project
- Source from namespace
- Source from commit

You'd end up with some redundancies with regard to fetching the build result, but that's already true, and I don't mind redundancy if it improves readability and thus maintainability.